urle381b6cba7_IkarosKappler_ikrs_http_tgz-pJ8-ikrs_json_rpc_JSONRPCHandlerJ8/src/ikrs/httpd/HTTPHandler.java

```diff
@@ -0,0 +1,1131 @@
+	
+
+	// Pre start core thread?
+	// this.executorService.prestartCoreThread();
+
+	logger.log( Level.INFO,
+		    getClass().getName() + "{init}",
+		    "Initialization done. System ready.");
+
+	// Store first instance
+	if( HTTPHandler.sharedInstance == null )
+	    HTTPHandler.sharedInstance = this;
+    }
+
+    private void initDefaultDocumentRoot() {
+	this.setDocumentRoot( new File("document_root") );
+
+    }
+
+    protected void setDocumentRoot( File newDocumentRoot ) 
+	throws NullPointerException {
+	if( newDocumentRoot == null ) {
+
+	    this.logger.log( Level.SEVERE,
+			     getClass().getName(),
+			     "[Init] Cannot set DOCUMENT_ROOT to null!" );
+	    throw new NullPointerException( "Cannot set DOCUMENT_ROOT to null!" );
+	}
+
+	this.documentRoot = newDocumentRoot; 
+	if( !this.documentRoot.exists() && !this.documentRoot.mkdirs() ) {
+
+	    this.logger.log( Level.SEVERE,
+			     getClass().getName(),
+			     "[Init] Failed to create document root '" + this.documentRoot.getPath() + "'! This will probably make your server un-usable. (continue though)" );
+
+	}
+    }
+
+    /**
+     * This method checks whether the passed file-system file is 'inside' the configured document root.
+     * The passed file should be rooted at the global file system root.
+     *
+     * Note that the document root directory is considered to be inside itself, as it is
+     * accessible by the URI '/'.
+     *
+     * @param file A file inside the local system's file system; must not be null.
+     * @return True if - and only if - the passed file is a file or a directory inside the document root
+     *              directory or inside any subdirectory inside the document root or the passed file
+     *              is document root itself. Or in other words: if the passed file is inside the configured 
+     *              HTTP document tree.
+     * @throws NullPointerException If the passed file is null.
+     **/ 
+    public boolean isInsideDocumentRoot( File file ) 
+	throws NullPointerException {
+
+	while( file != null ) {
+	    
+	    if( this.getDocumentRoot().equals(file) )
+		return true;
+
+	    file = file.getParentFile();
+	}
+
+	return false;
+    }
+    
+
+    /**
+     * This method initializes the FileHandler map.
+     **/
+    protected void initFileHandlers( File fileHandlersFile ) {
+
+	this.fileHandlerExtensionMap = Collections.synchronizedMap( new FileExtensionKeyMap<FileHandler>() );
+	this.fileHandlerNameMap      = Collections.synchronizedMap( new TreeMap<String,FileHandler>() );
+
+	//String fileHandlersFileName = "filehandlers.ini";
+	String handlerClassName = null; // "ikrs.http.filehandler.PHPHandler";
+	try {   
+
+	    //BufferedReader reader = new BufferedReader( new InputStreamReader( new FileInputStream(new File(fileHandlersFileName)) ) );
+	    BufferedReader reader = new BufferedReader( new InputStreamReader( new FileInputStream(fileHandlersFile) ) );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *reader*. Execution paths that do not contain closure statements were detected. Ensure *reader* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).